platforms/nuttx/src/px4/common/gpio/mcp23009/mcp23009.cpp

```diff
@@ -31,182 +31,122 @@
  *
  ****************************************************************************/
 
+#include <nuttx/ioexpander/gpio.h>
 #include <drivers/drv_sensor.h>
-#include <px4_platform/gpio/mcp23009.hpp>
-#include "mcp23009_registers.hpp"
-
-using namespace Microchip_MCP23009;
-
-const struct gpio_operations_s MCP23009::gpio_ops = {
-go_read : MCP23009::go_read,
-go_write : MCP23009::go_write,
-go_attach : nullptr,
-go_enable : nullptr,
-go_setpintype : MCP23009::go_setpintype,
+#include <lib/drivers/device/Device.hpp>
+#include <uORB/topics/gpio_config.h>
+#include <uORB/topics/gpio_in.h>
+#include <uORB/topics/gpio_out.h>
+#include <uORB/topics/gpio_request.h>
+#include <uORB/Publication.hpp>
+#include <uORB/SubscriptionBlocking.hpp>
+
+static uint32_t DEVID{0};
+struct mcp23009_gpio_dev_s {
+	struct gpio_dev_s gpio;
+	uint8_t mask;
 };
 
-MCP23009::MCP23009(int bus, int address, int first_minor, int bus_frequency) :
-	I2C(DRV_GPIO_DEVTYPE_MCP23009, "MCP23009", bus, address, bus_frequency),
-	_first_minor(first_minor)
-{
-}
-
-MCP23009::~MCP23009()
+static uORB::Publication<px4::msg::GpioRequest> toGpioRequest{ORB_ID(gpio_request)};
+static uORB::SubscriptionBlocking<px4::msg::GpioIn> fromGpioIn{ORB_ID(gpio_in)};
+static int mcp23009_read(struct gpio_dev_s *dev, bool *value)
 {
-	/* set all as input & unregister */
-	for (int i = 0; i < num_gpios; ++i) {
-		go_setpintype(i, GPIO_INPUT_PIN);
-		gpio_pin_unregister(&_gpio[i].gpio, _first_minor + i);
+	mcp23009_gpio_dev_s *gpio = (struct mcp23009_gpio_dev_s *)dev;
+	px4::msg::GpioRequest msg{
+		hrt_absolute_time(),
+		DEVID,
+	};
+	if (!toGpioRequest.publish(msg)) {
+		return -ETIMEDOUT;
 	}
-}
 
-int MCP23009::go_read(struct gpio_dev_s *dev, bool *value)
-{
-	mcp23009_gpio_dev_s *gpio = (struct mcp23009_gpio_dev_s *)dev;
-	return gpio->obj->go_read(gpio->id, value);
+	px4::msg::GpioIn input; // Wait up to 10ms
+	if (!fromGpioIn.updateBlocking(input, 10'000)) {
```

Comment:
This can block the calling work queue.

Is this wrapper going through nuttx files still required? Why not publish the uorb topics directly from adc?

Reply:
Due to the asynchronous uORB API, the read needs to wait *somewhere*, could be in the NuttX GPIO abstraction or directly in the ADC driver, but both locations will be in the hp_default work queue. Technically this was also true before, since the I2C bus transfer could fail and also takes some time, it was just hidden in the "blocking" API.

I can reduce this timeout to 1ms, but I'm not sure if there is a fundamentally good solution to GPIO expanders.
I kept the NuttX GPIO abstraction for backward compatibility and I didn't see any loops in the code where it automatically retries on failure, so it should "only" slow down the work queue, not block it entirely (not great though).

Reply:
Yes you need to wait but that does not mean it needs to be blocking. I.e. publish uorb request, continue, and check for a response in later iterations.

Reply:
ok, that's a good point! Then the GPIO abstraction needs to go away and the ADC calls the driver directly.